Purchases/Public/RCPurchases.m

```diff
@@ -163,88 +181,82 @@ - (void)addAttributionData:(NSDictionary *)data
-            }
+            return info;
         }
     }
+    return nil;
 }
 
 - (void)cachePurchaserInfo:(RCPurchaserInfo *)info {
-    if (info.JSONObject) {
-        NSError *jsonError = nil;
-        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:info.JSONObject
-                                                           options:0
-                                                             error:&jsonError];
-        if (jsonError == nil) {
-            [self.userDefaults setObject:jsonData
-                                  forKey:self.purchaserInfoUserDefaultCacheKey];
-        }
+    if (info) {
+        [self dispatch:^{
+            if (info.JSONObject) {
+                NSError *jsonError = nil;
+                NSData *jsonData = [NSJSONSerialization dataWithJSONObject:info.JSONObject
+                                                                   options:0
+                                                                     error:&jsonError];
+                if (jsonError == nil) {
+                    [self.userDefaults setObject:jsonData
+                                          forKey:self.purchaserInfoUserDefaultCacheKey];
+                }
+            }
+            self.cachesLastUpdated = [NSDate date];
+        }];
     }
 }
 
-- (void)updateCaches
-{
+- (BOOL)isCacheStale {
     NSTimeInterval timeSinceLastCheck = -[self.cachesLastUpdated timeIntervalSinceNow];
-    if (self.cachesLastUpdated != nil && timeSinceLastCheck < 60.) {
-        [self readPurchaserInfoFromCache];
-        return;
-    }
+    return (self.cachesLastUpdated != nil && timeSinceLastCheck < 60. * 5);
+}
 
-    self.cachesLastUpdated = [NSDate date];
+- (void)updateCaches {
+    [self updateCachesWithCompletionBlock:nil];
+}
 
+- (void)updateCachesWithCompletionBlock:(RCReceivePurchaserInfoBlock _Nullable)completion
+{
+    [self updatePurchaserInfoCache:completion];
+    [self updateEntitlementsCache:nil];
+}
+
+- (void)updatePurchaserInfoCache:(RCReceivePurchaserInfoBlock _Nullable)completion
+{
     [self.backend getSubscriberDataWithAppUserID:self.appUserID
                                       completion:^(RCPurchaserInfo * _Nullable info,
                                                    NSError * _Nullable error) {
-        if (error == nil) {
-            [self handleUpdatedPurchaserInfo:info error:nil];
-        } else {
-            self.cachesLastUpdated = nil;
+                                          if (error == nil) {
+                                              [self cachePurchaserInfo:info];
+                                          } else {
+                                              self.cachesLastUpdated = nil;
+                                          }
+                                          CALL_AND_DISPATCH_IF_SET(completion, info, error);
+                                      }];
+}
+
+- (void)purchaserInfoWithCompletionBlock:(RCReceivePurchaserInfoBlock)completion
```

Comment:
@vegaro The behavior changed here.

If we have an info in the cache, send it right away. Then, if the cache was stale, we fire off an update. If the one we fetch is updated, we'll trigger the delegate.

If we don't have an info in cache, just start an update.